core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -3443,12 +3444,32 @@ private RexNode convertUsing(SqlValidatorNamespace leftNamespace,
           throw new AssertionError("field " + name + " is not found in "
               + rowType + " with " + nameMatcher);
         }
+
+        fieldTypes.add(field.getType());
         operands.add(
             rexBuilder.makeInputRef(field.getType(),
                 offset + field.getIndex()));
         offset += rowType.getFieldList().size();
       }
-      list.add(rexBuilder.makeCall(SqlStdOperatorTable.EQUALS, operands));
+
+      RelDataType resultType =
+          validator().getTypeCoercion().commonTypeForBinaryComparison(
+              fieldTypes.get(0), fieldTypes.get(1));
+      if (resultType == null) {
+        throw new IllegalArgumentException("Cannot compute compatible row type: " + fieldTypes);
```

Comment:
`row type` or `row types`?

Reply:
I guess this exception message is not very good, I will change it.
What happens here is that the join is done on two fields which cannot be compared.
But I am not sure this exception can ever happen, because at this point the program is validated already, so if the fields are not comparable, this point should not be reached.

Reply:
Thanks for the clarify, I think you are right about whether exception can be thrown here.

Reply:
I will try to add a test to see what happens.

Reply:
I have checked, the validator will reject a program with a message like 
`Column 'X' matched using NATURAL keyword or USING clause has incompatible types: cannot compare`
I will add a comment that the exception is supposed to be unreachable.

Reply:
I have made the code clearer - but by mistake I amended the commit instead of adding a new one.

Reply:
Now the exception message is more clear.